File: 2016/Robot2016/src/com/kennedyrobotics/commands/SensorUpdateCommand.java
package com.kennedyrobotics.commands;

import com.kennedyrobotics.Robot;

import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 *
 */
public class SensorUpdateCommand extends Command {

	private boolean imuInit = false;
	
    public SensorUpdateCommand() {
    	requires(Robot.sensorSubsystem);
    }
    
    private double last_world_linear_accel_x;
    private double last_world_linear_accel_y;
    public final static double kCollisionThreshold_DeltaG = 0.5f;


    // Called just before this Command runs the first time
    protected void initialize() {
    	imuInit = false;
    	Robot.sensorSubsystem.setEdgeLed(true);
    }

    // Called repeatedly when this Command is scheduled to run
    protected void execute() {
    	if(!imuInit) {
			if(!Robot.sensorSubsystem.getNavX().isCalibrating()) {
				System.out.println("[SensorUpdateCommand] navX calibrated");
				imuInit = true;
			} else {
				//System.out.println("[SensorUpdateCommand] Waiting for navX to be calibrated");
			}
	    }
    	
    	boolean collisionDetected = false;
        
        double curr_world_linear_accel_x = Robot.sensorSubsystem.getNavX().getWorldLinearAccelX();
        double currentJerkX = curr_world_linear_accel_x - last_world_linear_accel_x;
        last_world_linear_accel_x = curr_world_linear_accel_x;
        double curr_world_linear_accel_y = Robot.sensorSubsystem.getNavX().getWorldLinearAccelY();
        double currentJerkY = curr_world_linear_accel_y - last_world_linear_accel_y;
        last_world_linear_accel_y = curr_world_linear_accel_y;
        
        if ( ( Math.abs(currentJerkX) > kCollisionThreshold_DeltaG ) ||
             ( Math.abs(currentJerkY) > kCollisionThreshold_DeltaG) ) {
            collisionDetected = true;
        }
        
        Robot.sensorSubsystem.setCollision(collisionDetected);
        SmartDashboard.putBoolean("Collision detected", collisionDetected);

        //Chasis
    	SmartDashboard.putNumber("Chassis Left Encoder", Robot.sensorSubsystem.getChassisLeftEncoder().getDistance());
    	SmartDashboard.putNumber("Chassis Right Encoder", Robot.sensorSubsystem.getChassisRightEncoder().getDistance());
    	
    	SmartDashboard.putNumber("Chassis Left Rate", Robot.sensorSubsystem.getChassisLeftEncoder().getRate());
    	SmartDashboard.putNumber("Chassis Right Rate", Robot.sensorSubsystem.getChassisRightEncoder().getRate());

    	SmartDashboard.putNumber("Chassis Left Encoder Samples", Robot.sensorSubsystem.getChassisLeftEncoder().getSamplesToAverage());
    	SmartDashboard.putNumber("Chassis Right Encoder Samples", Robot.sensorSubsystem.getChassisRightEncoder().getSamplesToAverage());
    	
    	//Arm
//    	SmartDashboard.putBoolean("Arm Aim Low Limit", Robot.sensorSubsystem.getArmTiltLowLimit());
//    	SmartDashboard.putBoolean("Arm Aim High Limit", Robot.sensorSubsystem.getArmTiltHighLimit());
    	//Shooter
    	SmartDashboard.putBoolean("Shooter Aim Low Limit", Robot.sensorSubsystem.getShooterAimLowLimit());
    	SmartDashboard.putBoolean("Shooter Aim High Limit", Robot.sensorSubsystem.getShooterAimHighLimit());
    	SmartDashboard.putNumber("Shooter Aim Encoder Position", Robot.sensorSubsystem.getShooterAimEncoder().getDistance());

    	SmartDashboard.putBoolean("Line Dedector", Robot.sensorSubsystem.getLineDedector());;
    	
    	//Nav X
    	SmartDashboard.putBoolean("NavX Connected", Robot.sensorSubsystem.getNavX().isConnected());
    	SmartDashboard.putBoolean("NavX Calibrating", Robot.sensorSubsystem.getNavX().isCalibrating());
    	SmartDashboard.putNumber("NavX Yaw", Robot.sensorSubsystem.getNavX().getYaw());
    	SmartDashboard.putNumber("NavX Pitch", Robot.sensorSubsystem.getNavX().getPitch());
    	SmartDashboard.putNumber("NavX Roll", Robot.sensorSubsystem.getNavX().getRoll());
    	SmartDashboard.putNumber("NavX Compas Heading", Robot.sensorSubsystem.getNavX().getCompassHeading());
    	SmartDashboard.putNumber("Yaw Angle", Robot.sensorSubsystem.getYaw());
    	
    }

    // Make this return true when this Command no longer needs to run execute()
    protected boolean isFinished() {
        return false;
    }

    // Called once after isFinished returns true
    protected void end() {
    }

    // Called when another command which requires one or more of the same
    // subsystems is scheduled to run
    protected void interrupted() {
    }
}

File: 2017/2017Robot/src/Commands/SensorUpdateCommand.cpp
#include "SensorUpdateCommand.h"
#include <SmartDashboard/SmartDashboard.h>
#include "../RobotMap.h"

SensorUpdateCommand::SensorUpdateCommand() {
	// Use Requires() here to declare subsystem dependencies
	// eg. Requires(Robot::chassis.get());
	Requires(sensorSubsystem.get());
}

// Called just before this Command runs the first time
void SensorUpdateCommand::Initialize() {
}

// Called repeatedly when this Command is scheduled to run
void SensorUpdateCommand::Execute() {

	SmartDashboard::PutNumber("Chassis Left Distance", sensorSubsystem->GetChassisLeftEncoder()->GetDistance());
	SmartDashboard::PutNumber("Chassis Right Distance", sensorSubsystem->GetChassisRightEncoder()->GetDistance());

	SmartDashboard::PutNumber("Chassis Left Raw", sensorSubsystem->GetChassisLeftEncoder()->GetRaw());
	SmartDashboard::PutNumber("Chassis Right Raw", sensorSubsystem->GetChassisRightEncoder()->GetRaw());

	SmartDashboard::PutNumber("Chassis Left Rate",  sensorSubsystem->GetChassisLeftEncoder()->GetRate());
	SmartDashboard::PutNumber("Chassis Right Rate", sensorSubsystem->GetChassisRightEncoder()->GetRate());

	//SmartDashboard::PutNumber("Left Out", chassis->GetLeftMaster()->Get());
	SmartDashboard::PutNumber("Right Out", chassis->GetRightMaster()->Get());

//	SmartDashboard::PutNumber("Chassis Left Current", sensorSubsystem->GetChassisLeftCurrent());
//	SmartDashboard::PutNumber("Chassis Right Current", sensorSubsystem->GetChassisRightCurrent());
	SmartDashboard::PutNumber("Left Master Voltage", chassis->GetLeftMaster()->GetOutputVoltage());
	SmartDashboard::PutNumber("Right Master Voltage", chassis->GetRightMaster()->GetOutputVoltage());
	SmartDashboard::PutNumber("Left Slave Voltage", chassis->GetLeftSlave()->GetOutputVoltage());
	SmartDashboard::PutNumber("Right Slave Voltage", chassis->GetRightSlave()->GetOutputVoltage());

	SmartDashboard::PutNumber("Left Master Current", chassis->GetLeftMaster()->GetOutputCurrent());
	SmartDashboard::PutNumber("Right Master Current", chassis->GetRightMaster()->GetOutputCurrent());
	SmartDashboard::PutNumber("Left Slave Current", chassis->GetLeftSlave()->GetOutputCurrent());
	SmartDashboard::PutNumber("Right Slave Current", chassis->GetRightSlave()->GetOutputCurrent());

	SmartDashboard::PutNumber("Climber 0 Current", sensorSubsystem->GetPDP()->GetCurrent(PDP_CLIMBER_0));
	SmartDashboard::PutNumber("Climber 1 Current", sensorSubsystem->GetPDP()->GetCurrent(PDP_CLIMBER_1));

	SmartDashboard::PutBoolean("Climber At Index", sensorSubsystem->GetClimberIndex());

	auto imu = sensorSubsystem->GetIMU();
	if (NULL != imu)
	{
		SmartDashboard::PutNumber("IMU Yaw", sensorSubsystem->GetIMU()->GetYaw());
		SmartDashboard::PutNumber("IMU Rate", sensorSubsystem->GetIMU()->GetRate());
		SmartDashboard::PutBoolean("IMU Connected", sensorSubsystem->GetIMU()->IsConnected());
		SmartDashboard::PutBoolean("IMU Calibrating", sensorSubsystem->GetIMU()->IsCalibrating());
	}

	SmartDashboard::PutBoolean("A", oi->GetRevDigit()->GetA());
	SmartDashboard::PutBoolean("B", oi->GetRevDigit()->GetB());
	SmartDashboard::PutNumber("Pot", oi->GetRevDigit()->GetPotVoltage());


	SmartDashboard::PutNumber("Yaw", sensorSubsystem->GetYaw());
}

// Make this return true when this Command no longer needs to run execute()
bool SensorUpdateCommand::IsFinished() {
	return false;
}

// Called once after isFinished returns true
void SensorUpdateCommand::End() {

}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void SensorUpdateCommand::Interrupted() {

}
